pkg/kubelet/nodeconfig/controller.go

```diff
@@ -0,0 +1,369 @@
+			return
+		}
+	} else if !cc.curIsLkg() {
+		// when the trial period is over, the current config becomes the last-known-good
+		cc.setCurAsLkg()
+	}
+
+	// update the status to note that we will use the current config
+	cc.setConfigOK(fmt.Sprintf(curRemoteMessageFmt, curUID), curRemoteOKReason, apiv1.ConditionTrue)
+	finalConfig = cur
+	return
+}
+
+// StartSyncLoop launches the sync loop that watches the API server for configuration updates
+// The `client` passed to StartSyncLoop will replace the controller's current client
+func (cc *NodeConfigController) StartSyncLoop(client clientset.Interface, nodeName string) {
+	if client != nil {
+		cc.client = client
+		cc.nodeName = nodeName
+		infof("starting sync loop")
+
+		// start the configuration sync loop
+		go func() {
+			defer utilruntime.HandleCrash()
+			fieldselector := fmt.Sprintf("metadata.name=%s", cc.nodeName)
+
+			// Add some randomness to resync period, which can help avoid controllers falling into lock-step
+			minResyncPeriod := 30 * time.Second
+			factor := rand.Float64() + 1
+			resyncPeriod := time.Duration(float64(minResyncPeriod.Nanoseconds()) * factor)
+
+			lw := &cache.ListWatch{
+				ListFunc: func(options metav1.ListOptions) (kuberuntime.Object, error) {
+					return cc.client.Core().Nodes().List(metav1.ListOptions{
+						FieldSelector: fieldselector,
+					})
+				},
+				WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+					return cc.client.Core().Nodes().Watch(metav1.ListOptions{
+						FieldSelector:   fieldselector,
+						ResourceVersion: options.ResourceVersion,
+					})
+				},
+			}
+
+			handler := cache.ResourceEventHandlerFuncs{
+				AddFunc: func(obj interface{}) {
+					cc.onAddNodeEvent(obj)
+				},
+				UpdateFunc: func(old interface{}, nw interface{}) {
+					cc.onWatchNodeEvent(nw)
+				},
+				DeleteFunc: func(obj interface{}) {
+					cc.onDeleteNodeEvent(obj)
+				},
+			}
+
+			informer := cache.NewSharedInformer(lw, &apiv1.Node{}, resyncPeriod)
+			informer.AddEventHandler(handler)
+			stop := make(chan struct{})
+			informer.Run(stop)
+			return
+		}()
+
+		// start the ConfigOK condition sync loop
+		// TODO(mtaufen): condition syncing is kept separate from informer event handling, because I'm worried about the
+		// possibility of creating a feedback loop between condition updates and Node update events.
+		// I have yet to test whether a feedback loop is possible, but it's best to be cautious.
+		go func() {
+			// Add some randomness to resync period, which can help avoid controllers falling into lock-step
+			minResyncPeriod := 10 * time.Second
+			factor := rand.Float64() + 1
+			resyncPeriod := time.Duration(float64(minResyncPeriod.Nanoseconds()) * factor)
+			// sync immediately, then periodically sync the ConfigOK condition
+			cc.syncConfigOK()
+			for {
+				select {
+				case <-time.After(resyncPeriod):
+					cc.syncConfigOK()
```

Comment:
This is going to increase the number of updates on the node object a lot. cc @gmarek